### ISSUE TITLE
Design of a simple fixed point library

### ISSUE THREAD

**Issue by jlperla** [JuliaNLSolvers/NLsolve.jl#144]
It might make sense to have it in NLsolve.jl, or it might make sense to have a FixedPoint.jl separately.  Regardless, here is a rough example of the simplest algorithm I was thinking of.  Two versions, one inplace and one outofplace?
```
function fixedpoint(f, x0; residualnorm = (x -> norm(x,Inf)), tol = 1E-10, maxiter=100)    
    residual = Inf
    iter = 1
    xold = x0
    while residual > tol && iter < maxiter
        xnew = f(xold)        
        residual = residualnorm(xold - xnew);
        xold = xnew
        iter += 1
    end
    return xold
end

function fixedpoint!(f!, x0; residualnorm = (x -> norm(x,Inf)), tol = 1E-10, maxiter=100)    
    residual = Inf
    iter = 1
    xold = x0
    xnew = copy(x0)
    while residual > tol && iter < maxiter
        f!(xold, xnew)        
        residual = residualnorm(xold - xnew);
        xold = xnew
        iter += 1
    end
    return xold
end
```

**Comment by jlperla** [comment]
Ideally, it would be great to be able to swap out algorithms.  I wrote down the following sketch of type dispatching, but not sure it is the best design
```
# Tags for algorithms.  Right now only iteration is there.
abstract type FixedPointAlgorithm end
type FixedPointIteration <: FixedPointAlgorithm end

function fixedpoint(x, algorithm::FixedPointIteration, tol)
    return 0 #implementation
end

function fixedpoint(x; algorithm::T = FixedPointIteration(), tol = 1E-10) where  {T <: FixedPointAlgorithm}
    fixedpoint(x, algorithm, tol)
end
fixedpoint(0)
```

**Comment by jlperla** [comment]
Finally, to ensure that people do not use the inplace version `fixedpoint!` with incompatible types, the initial value for the iteration could use a test like `x0 isa Number || x0 isa StaticArray`

**Comment by jlperla** [comment]
It would be nice for Newton iteration to allow passing in a jacobian, or using auto-differentiation of some sort otherwise.

**Comment by pkofod** [comment]
> Finally, to ensure that people do not use the inplace version fixedpoint! with incompatible types, the initial value for the iteration could use a test like x0 isa Number || x0 isa StaticArray

Hopefully, this will be possible to do using multiple dispatch soon after v1.0 is released and some sort of trait interface is developed for this.




> It would be nice for Newton iteration to allow passing in a jacobian, or using auto-differentiation of some sort otherwise.

this is already possible through `nlsolve` 



> Two versions, one inplace and one outofplace?

this is interesting for NLsolve and Optim as well, I'm stealing this idea and opening a new issue

**Comment by jlperla** [comment]
All sounds great.  The trait tests could be implemented later.  Do you think a new package should be created, or that this all should be part of `NLsolve.jl`?

**Comment by pkofod** [comment]
When Anderson's method was added it seemed like consensus was to keep fixed point solvers here, so I think it's fine to add them here. We can always move them out, but keeping it here makes for one less package to stumble upon for the users :)

**Comment by jlperla** [comment]
Fine with me.  As long as there is a convenient interface for getting a fixed point, I don't care where it lives.  The only concern is for less technical users, is setting up the `NLsolve.jl` development and test environment on their machines tricky?

**Comment by ChrisRackauckas** [comment]
It doesn't need all of the extra stuff. `fixedpoint` just needs to call `nlsolve` on `(x)->f(x)-x` and define the Jacobian to be the Jacobian of `f` and subtract 1 on the diagonals. It would just be a 2 line convenience function for any algorithm that's not Anderson acceleration. That way you get all of the other methods for free (like Newton trust-region, etc.) and there's no performance problem with doing this. In theory the user could always do this, but it's just a convenience function.

I don't see it necessary to make it another package since it's just 4-5 lines of code. If you want, a Picard method could be added here too.

>The only concern is for less technical users, is setting up the NLsolve.jl development and test environment on their machines tricky?

NLsolve.jl has no binary dependencies. Anywhere that Julia runs should work. 

>Finally, to ensure that people do not use the inplace version fixedpoint! with incompatible types, the initial value for the iteration could use a test like x0 isa Number || x0 isa StaticArray

Those aren't all of the types. What you're looking for instead is a test of mutability, which isn't robust. I would just let it run to method error since this is not something you can guess.

>Two versions, one inplace and one outofplace?

You can steal DiffEq's method for function identification. In theory there are ways to break it, but in practice it seems to work for everyone.

**Comment by jlperla** [comment]
@ChrisRackauckas  Whatever algorithm works, I don't especially care.

I don't think that calling `(x) -> f(x) - x` is the right thing to do in many circumstances.  With the sorts of fixed point problems you get in economics, there are convergence questions for solving a system of equations, while doing picard iteration style can be proven to converge.  But I agree that it would be really nice to allow it to use `nlsolve` in that way if possible and that it is a flag that would be nice to use.  I want economists to start using solvers for this sort of thing

However, it is possible that newton's fixed point iteration method would be similar to just using nlsolve on `(x)->f(x)-x` with newton's method... but I am not 100% sure of that, and certainly it would require a special case in a wrapper since the jacobian would be written for the $f(x)$ rather than $f(x) - x$

Sounds like between your function identification and the NLsolve ways for dealing with AD and optional jacobians, that the interface can come together easily.

All I want to make sure is that I can tell people to define an operator `F` and an initial condition and say `fixedpoint(F, x0)`... after that, helping them to try different algorithms is exactly what I was hoping for (especially with AD)

**Comment by pkofod** [comment]
> You can steal DiffEq's method for function identification. In theory there are ways to break it, but in practice it seems to work for everyone.

Maybe I should take a stab at that soon, I thought it would be fragile, but it never seems to come up in your ecosystem, so why should it here.

Regarding Picard, that is already in NLsolve. Actually, it's the standard when you choose Anderson, as it defaults to memoryless.



> but I am not 100% sure of that, and certainly it would require a special case in a wrapper since the jacobian would be written for the $f(x)$ rather than $f(x) - x$

Chris already mentioned that all you need to do is to subtract `I` in that case :) (we would do it through the convenience wrapper)

**Comment by pkofod** [comment]
> All I want to make sure is that I can tell people to define an operator F and an initial condition and say fixedpoint(F, x0)... after that, helping them to try different algorithms is exactly what I was hoping for (especially with AD)

This could be ready tomorrow :) (I don't have time right now, but... in less than a days work)

**Comment by ChrisRackauckas** [comment]
Basically DiffEq's route is this:

There's a type parameter on the problem types, like `ODEProblem{true}(f,u0,tspan)` is inplace, and false for out of place. Then there this code:

https://github.com/JuliaDiffEq/DiffEqBase.jl/blob/master/src/utils.jl#L1-L31

where `isinplace(f,k)` checks to see if there's a method with maximum arguments of greater than or equal to than `k` and if so then returns true, so for `f(du,u,p,t)` we have `k=4` and it'll be true and that's how the parameter is set automatically (it has extra tie-ins for DiffEq-specific Jacobian stuff, ignore that). 

There is one case this technically fails, and that's if you define a second dispatch which has a lot of parameters but don't want to set the inplaceness. In practice, that never actually happens.

You can make `nlsolve` etc. have that type parameter if you want by rewriting `nlsolve` to be a singleton where the inner contructor has a parameter and does exactly what it does now, and make an outer constructor just attach that parameter. Inner constructors don't actually have to return the type, so this is a little bit of a hack but it works.

Examples in action:

https://github.com/JuliaDiffEq/DiffEqBase.jl/blob/master/src/problems/ode_problems.jl#L4-L34

It's pretty straightforward once you know how to read from the methods table.

**Comment by ChrisRackauckas** [comment]
>I don't think that calling (x) -> f(x) - x is the right thing to do in many circumstances. With the sorts of fixed point problems you get in economics, there are convergence questions for solving a system of equations, while doing picard iteration style can be proven to converge. But I agree that it would be really nice to allow it to use nlsolve in that way if possible and that it is a flag that would be nice to use. I want economists to start using solvers for this sort of thing

Yes, you just subtract `I` from the Jacobian and its the same algorithm. Also, if you do Picard on `x->f(x)+x` then that's a rootfinding algorithm. The two problems are mathematically equivalent under this transformation, so you can mix-and-match the two ideas whenever necessary (as long as you account for the `I` in the Jacobian).

**Comment by jlperla** [comment]
Thanks to you both.  I love the idea of just telling economists to use the library for all fixed point problems (of which there are many...)

@pkofod Is there value in me getting an undergraduate to work with you on this for setting up tests, etc. or is it too much work to teach them how to setup the development and testing environment?

**Comment by pkofod** [comment]
> @pkofod Is there value in me getting an undergraduate to work with you on this for setting up tests, etc. or is it too much work to teach them how to setup the development and testing environment?

I'd be happy to help. It's not too difficult, really. It's surprisingly easy to transition from user to developer in Julia!

**Comment by jlperla** [comment]
Great, I may put an undergraduate in touch with you for getting setup with this for getting tests up and running.  The lack of binary dependencies will help a lot.

Two more things:
1. There are a lot of algorithms in economics where you apply a fixed point a set number of times (or at least not until convergence), do something else, then repeat the process.  This comes up solving dynamic programming problems since it is well known that policy functions converge much faster than value functions.
As an example, see modified policy function iteration in http://www.wiwi.uni-bonn.de/kuhn/notes/DPNotes.pdf

The basic approach is to use a value function to find an optimal policy, then fix the optimal policy and do value function iteration a bunch of times, solve the optimal policy again, etc.

It would be nice to be able to cleanly do that with the fixed point function.  Maybe the answer is just to tell people to set the number of iterations to a small number, but wanted to bring it up?

2. My original thought was that (at least the picard) fixedpoint iteration could be used for any julia type which has addition and a norm (maybe a Banach space or something like that?).  For example, you could find the fixed point of an operator on polynomials where the norm is the L2 norm or whatever.  Is this something to consider, and would it mess with the interfaces at all?

**Comment by pkofod** [comment]
> nown that policy functions converge much faster than value functions.

indeed, as it is equivalent to Newton iterations, it does. We could handle this via callbacks I guess. Currently, there are no callbacks in NLsolve, but I figure they could be useful in a lot of cases.



> My original thought was that (at least the picard) fixedpoint iteration could be used for any julia type which has addition and a norm (maybe a Banach space or something like that?). For example, you could find the fixed point of an operator on polynomials where the norm is the L2 norm or whatever. Is this something to consider, and would it mess with the interfaces at all?

Internally this would be quite different from how we usually handle arrays efficiently, but I'm all for allowing people to use these methods in their more general formulations insofar we can handle it in a somewhat coherent way.

**Comment by jlperla** [comment]
I think for now it might make sense to just plan on instructing people to pass in a small number for max iterations, but it is something to think about.

Lets just keep iterations on function spaces in mind for keeping the interface relatively general and maybe the future (i.e. thinking about using this sort of thing for collocation methods).

@jasmineyang is the undergraduate student who I would like to help out with setting up some tests, etc.  I will get point her in the right direction and ask her to get NLsolve.jl up and running on her computer.

**Comment by jlperla** [comment]
Closing this issue as sub-issues have been generated for the new branch.
